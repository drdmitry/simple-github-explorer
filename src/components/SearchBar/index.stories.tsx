import type { Meta, StoryObj } from '@storybook/react';

import { SearchBar } from './index';

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  title: 'Header/SearchBar',
  component: SearchBar,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: 'padded',
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ['autodocs'],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
  },
} satisfies Meta<typeof SearchBar>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
  },
  render: (args) => (
    <div className="max-w-lg mx-auto">
      <SearchBar {...args} />
    </div>
  ),
};

export const FullWidth: Story = {
  args: {
  },
  render: (args) => <SearchBar {...args} />,
};
