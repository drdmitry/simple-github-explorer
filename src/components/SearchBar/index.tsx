import { useState, type FC, useRef } from "react";
import { z, type ZodError } from "zod";

import Button from "~/components/Button";
import SvgIcon from "~/components/Icons/SvgIcon";

import useStore from "~/store/useStore";

const usernameSchema = z
  .string()
  .max(39, { message: "Username must be 39 characters or less" })
  .regex(/^[a-zA-Z\d](?:[a-zA-Z\d]|-(?=[a-zA-Z\d]))*$/, {
    message:
    "Username can only contain alphanumeric characters and hyphens, and it cannot start or end with a hyphen",
  });

export const SearchBar: FC = () => {
  const [searchTerm, setSearchTerm] = useState<string>("");
  const [validationError, setValidationError] = useState<string | null>(null);
  const inputRef = useRef<HTMLInputElement>(null);

  const {
    setSearchTerm: setStoreSearchTerm,
    searchUsers,
  } = useStore();

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
    try {
      if (event.target.value) usernameSchema.parse(event.target.value);
      setValidationError(null);
    } catch (error) {
      const errors = (error as ZodError)?.errors;
      if (errors && errors.length > 0 && errors[0]) {
        setValidationError(errors[0].message);
      }
    }
  };

  const handleInputContainerClick = () => {
    inputRef.current?.focus();
  };

  const hadleSearchClick = () => {
    setStoreSearchTerm(searchTerm);
    void searchUsers(searchTerm);
  };

  const handleKeyDown = (event: React.KeyboardEvent) => {
    if (event.key === 'Enter' || event.key === 'NumpadEnter') {
      event.preventDefault();
      setStoreSearchTerm(searchTerm);
      void searchUsers(searchTerm);
    } else if (event.key === 'Escape') {
      setSearchTerm("");
      setValidationError(null);
    }
  };

  const buttonDisabled = String(searchTerm).length === 0;

  return (
    <div className="w-full flex flex-col sm:flex-row gap-1">
      <div className="relative w-full">
        <div className="absolute h-10 w-10 px-2 flex items-center" onClick={handleInputContainerClick}>
          <SvgIcon name="search" size={24} className="text-gray-600" />
        </div>
        <input
          ref={inputRef}
          placeholder="Search username"
          autoComplete="off"
          autoCorrect="off"
          autoCapitalize="off"
          spellCheck="false"
          className="bg-gray-100 rounded-md border border-gray-300 py-2 px-4 pl-10 block w-full outline-none focus:shadow hover:border-gray-400 focus:border:gray-400"
          type="text"
          value={searchTerm}
          onChange={handleInputChange}
          onKeyDown={handleKeyDown}
        />
        {validationError && (
          <div className="absolute px-1 text-xs text-red-400">
            {validationError}
          </div>
        )}
      </div>
      <Button
        disabled={buttonDisabled}
        onClick={hadleSearchClick}
      >
        Search
      </Button>
    </div>
  );
};

export default SearchBar;
