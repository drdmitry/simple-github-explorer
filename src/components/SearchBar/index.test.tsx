import React from "react";
import { render, fireEvent } from "@testing-library/react";

import SearchBar from "./";

describe("SearchBar", () => {
  test("renders SearchBar", () => {
    // Render the SearchBar component
    const { getByText, getByPlaceholderText, container } = render(
      <SearchBar />
    );
    expect(container).toBeInTheDocument();

    const svgElement = container.querySelector("svg");
    expect(svgElement).toBeInTheDocument();

    const inputElement = getByPlaceholderText("Search username");
    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toBeInstanceOf(HTMLInputElement);
    expect(inputElement).toHaveValue('');

    // Assert that the Search Button is rendered
    const buttonElement = getByText("Search");
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toBeInstanceOf(HTMLButtonElement);
  });

  test('handles input change and validation', () => {
    const { getByPlaceholderText, getByText } = render(<SearchBar />);
    const inputElement = getByPlaceholderText('Search username');

    // Simulate valid input
    fireEvent.change(inputElement, { target: { value: 'valid-username' } });
    expect(inputElement).toHaveValue('valid-username');

    // Simulate invalid input
    fireEvent.change(inputElement, { target: { value: '-invalid-username' } });
    expect(inputElement).toHaveValue('-invalid-username');
    expect(getByText('Username can only contain alphanumeric characters and hyphens, and it cannot start or end with a hyphen')).toBeInTheDocument();
  });

  test("renders SearchBar unchanged", () => {
    const { container } = render(<SearchBar />);
    expect(container).toMatchSnapshot();
  });
});
