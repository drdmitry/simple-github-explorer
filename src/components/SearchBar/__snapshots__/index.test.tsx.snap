// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar renders SearchBar unchanged 1`] = `
<div>
  <div
    class="w-full flex flex-col sm:flex-row gap-1"
  >
    <div
      class="relative w-full"
    >
      <div
        class="absolute h-10 w-10 px-2 flex items-center"
      >
        <svg
          class="text-gray-600"
          fill="none"
          height="24"
          stroke="currentColor"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1"
          />
        </svg>
      </div>
      <input
        autocapitalize="off"
        autocomplete="off"
        autocorrect="off"
        class="bg-gray-100 rounded-md border border-gray-300 py-2 px-4 pl-10 block w-full outline-none focus:shadow hover:border-gray-400 focus:border:gray-400"
        placeholder="Search username"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
    <button
      class="bg-sky-500 bg-gradient-to-br from-sky-300 to-sky-500 hover:to-sky-600 active:to-sky-700 text-white shadow-sm font-bold py-2 px-4 rounded-md focus:outline-none focus:shadow-outline disabled:from-sky-100 disabled:to-sky-200 disabled:cursor-no-drop"
      disabled=""
      type="button"
    >
      Search
    </button>
  </div>
</div>
`;
